CruzHacks/cruzhacks-2021-website
# Enforce Branch Naming Conventions

Problem
=======
As a developer
I want to enforce branch naming conventions 
So our CI/CD pipeline can run smoothly

[Link to Asana Ticket](https://app.asana.com/0/1197738625416321/1199386537603653/f)

Solution
========
**Note**: *I followed a couple of articles that helped with this.*

Git provides something really neat called [*GitHooks*](https://githooks.com/). Each git repo actually comes with a hooks directory under the `.git` folder, which never gets checked into version control. I created a pre-commit script, which runs when you commit changes into a repository. 

The following naming conventions will be followed within this project: 
* `feature/` 
    * For any big feature stories such as the creation of a component. Pretty much all the work we will be doing will fall under feature
* `bug/`
    * If we find a bug within the site, we can file a bug ticket in Asana and create a fix using this branch naming convention
* `spike/`
    * if we are spiking a new tool or technology, it can go under this branch naming convention 
* `deploy/`
    * These branches will be reserved for deploys to production only 

These branch naming conventions will help a ton when creating the Changelog for deploy PR's. We can use the branch names to categorize the commits. [This article](https://linuxize.com/post/how-to-rename-local-and-remote-git-branch/) does a really good job in walking through how to rename branches if they are not adhering to the branch naming conventions. 

Credits to [this article](https://itnext.io/using-git-hooks-to-enforce-branch-naming-policy-ffd81fa01e5e)

Change summary:
---------------
* added pre-commit error message for branch naming conventions c0a7575
* added executable file 7a7ef43

Steps to Verify:
----------------
1. Checkout this branch and run the following command at the root of the directory: `git config core.hooksPath .githooks`
   * this will configure your hooks path to point to the directory we are pushing in here 
1. Within this same branch, run the following command: `git checkout -b testNamingConvention` 
1. Change any line in the a file within the repository and commit the change. Verify that the error message shows up. 

Screenshots:
-----------------------
**Error Message**
<img width="1101" alt="Screen Shot 2020-11-29 at 11 07 17 PM" src="https://user-images.githubusercontent.com/40477553/100578675-a3884780-3297-11eb-82d2-51a2908da71f.png">
